backend-comparison/src/burnbenchapp/progressbar.rs

```diff
@@ -0,0 +1,118 @@
+use core::fmt;
+use std::{
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+
+use indicatif::{style::ProgressTracker, ProgressBar, ProgressState, ProgressStyle};
+
+pub(crate) struct RunnerProgressBar {
+    pb: ProgressBar,
+    tracker: ThreadSafeTracker,
+}
+
+impl RunnerProgressBar {
+    pub(crate) fn new(total: u64) -> Self {
+        let tracker = CountTracker::default();
+        let thread_safe_tracker = ThreadSafeTracker::new(tracker);
+        let pb = ProgressBar::new(total);
+        pb.set_style(
+            ProgressStyle::default_spinner()
+                .template("\n{msg}\n{spinner}{wide_bar:.yellow/red} {pos}/{len} {counter}\n ")
+                .unwrap()
+                .with_key("counter", thread_safe_tracker.clone())
+                .progress_chars("▬▬―")
+                .tick_strings(&[
+                    "🕛 ", "🕐 ", "🕑 ", "🕒 ", "🕓 ", "🕔 ", "🕕 ", "🕖 ", "🕗 ", "🕘 ", "🕙 ",
+                    "🕚 ",
+                ]),
+        );
+        Self {
+            pb,
+            tracker: thread_safe_tracker,
+        }
+    }
+
+    pub(crate) fn message(&self, msg: String) {
+        self.pb.set_message(msg);
+    }
+
+    pub(crate) fn advance_spinner(&self) {
+        self.pb.tick();
+    }
+
+    /// make the spinner to spin automatically
+    pub(crate) fn start_spinner(&self) {
+        self.pb.enable_steady_tick(Duration::from_millis(100));
+    }
+
+    /// stop the spinner to spin automatically
+    pub(crate) fn stop_spinner(&self) {
+        self.pb.disable_steady_tick();
+    }
+
+    pub(crate) fn inc_by_one(&self) {
+        self.pb.inc(1);
+    }
+
+    pub(crate) fn succeeded_inc(&mut self) {
+        self.tracker.inner.lock().unwrap().succeeded += 1;
+    }
+
+    pub(crate) fn failed_inc(&mut self) {
+        self.tracker.inner.lock().unwrap().failed += 1;
+    }
+
+    pub(crate) fn finish(&self) {
+        let success = self.tracker.inner.lock().unwrap().failed == 0;
+        let msg = format!(
+            "\n{{msg}}\n{{wide_bar:.{}}}",
+            if success { "green" } else { "red" }
+        );
+        self.pb.set_style(
+            ProgressStyle::with_template(&msg)
+                .unwrap()
+                .progress_chars("▬▬―"),
+        );
+        self.pb.finish_with_message(if success {
+            "Benchmarks Complete!"
+        } else {
+            "Some benchmarks failed!"
+        });
+    }
+}
+
+#[derive(Clone, Default)]
+struct CountTracker {
+    pub failed: u64,
+    pub succeeded: u64,
+}
+
+#[derive(Clone)]
+struct ThreadSafeTracker {
+    inner: Arc<Mutex<CountTracker>>,
+}
```

Comment:
That's a bit more complicated than that but maybe I am mistaken.

I need to pass the tracker down to progress bar style but I also need to pass it to my runner to it can update its state. I tried with an Arc like that:

```rust
pub(crate) struct RunnerProgressBar {
    pb: ProgressBar,
    tracker: Arc<Mutex<CountTracker>>,
}

impl RunnerProgressBar {
    pub(crate) fn new(total: u64) -> Self {
        let tracker = Arc::new(Mutex::new(CountTracker::default()));
        let pb = ProgressBar::new(total);
        pb.set_style(
            ProgressStyle::default_spinner()
                .template("\n{msg}\n{spinner}{wide_bar:.yellow/red} {pos}/{len} {counter}\n ")
                .unwrap()
                .with_key("counter", tracker.clone())
                .progress_chars("▬▬―")
                .tick_strings(&[
                    "🕛 ", "🕐 ", "🕑 ", "🕒 ", "🕓 ", "🕔 ", "🕕 ", "🕖 ", "🕗 ", "🕘 ", "🕙 ",
                    "🕚 ",
                ]),
        );
        Self {
            pb,
            tracker: tracker.clone(),
        }
    }
```

But then the Tracker trait is not implemented for the Arc'd CountTracker type and did not find an easy way to do it so I wrapped it.

Thinking about it, maybe I can avoid owning the tracker at the runner level if the indicatif interface gives me access to the tracker somehow. I'll take a look at this.

Reply:
I refactored the tracker and use a ref counted Atomic64. Pretty happy about it, thank you for the suggestion. It is much cleaner. Note that I used the weakest form of ordering as I am not familiar with the concept yet and atomicity only seems to just fined.